src/control/positionControl/BBTrajectories/WorldObjects.cpp

```diff
@@ -0,0 +1,183 @@
+        }
+    }
+
+    void
+    WorldObjects::calculateEnemyRobotCollisions(rtt::BB::BBTrajectory2D BBTrajectory, std::vector<CollisionData> &collisionDatas,
+                                                std::vector<double> &collisionTimes, const std::vector<Vector2> &pathPoints,
+                                                double timeStep) {
+        auto theirRobots = world->getWorld()->getThem();
+
+        for (int j = 0; j < pathPoints.size(); j++) {
+            for (int i = 0; i < theirRobots.size() - 1; i++) {
+                double currentTime = j * timeStep;
+                double maxCollisionCheckTime = 2;
+                if (currentTime <= maxCollisionCheckTime) {
+                    // TODO: Currently enemy position in future is calculated with simple model (x = x + v*t), maybe improve?
+                    Vector2 theirVel = theirRobots[i]->getVel();
+                    Vector2 theirPos = theirRobots[i]->getPos() /*+ theirVel * currentTime*/;
+                    Vector2 posDif = BBTrajectory.getPosition(currentTime) -
+                                     theirPos; //TODO: Is currentTime defined the same way as the time from the BBTrajectory time?
+                    if (posDif.length() < 3 /*<-- kawaiii*/ * ai::Constants::ROBOT_RADIUS_MAX()) {
+                        Vector2 ourVel = BBTrajectory.getVelocity(currentTime);
+                        Vector2 velDif = ourVel - theirVel;
+                        double projectLength = velDif.dot(posDif) / sqrt(posDif.dot(posDif));
+                        if (abs(projectLength) > 1.5 && theirVel.length() < ourVel.length()) {
+                            collisionDatas.emplace_back(CollisionData{theirPos, pathPoints[j]});
+                            collisionTimes.emplace_back(currentTime);
+                            return;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    void
+    WorldObjects::calculateOurRobotCollisions(std::vector<CollisionData> &collisionDatas, std::vector<double> &collisionTimes,
+                                              const std::vector<Vector2> &pathPoints, int robotId, double timeStep) {
+
+        int ourRobotAmount = world->getWorld()->getUs().size();
+
+        for (int j = 0; j < pathPoints.size(); j++) {
+            for (int i = 0; i < ourRobotAmount; i++) {
+                if (robotId != i && !calculatedPaths[i].empty()) {
+                    if ((pathPoints[j] - calculatedPaths[i][j]).length() < ai::Constants::ROBOT_RADIUS() * 1.5 &&
+                        j * timeStep < 1) {
+                        collisionDatas.emplace_back(CollisionData{calculatedPaths[i][j], pathPoints[j]});
+                        collisionTimes.emplace_back(j * timeStep);
+                        return;
+                    }
+                }
+            }
+        }
+    }
+
+
+    void WorldObjects::setField(const rtt::ai::rtt_world::Field &field_) { this->field = &field_; }
```

Comment:
This is very dangerous; if the field pointer in world changes for whatever reason, you're going to get a segfault at some point in this class because you copied it. This could happen in many circumstances. I'd personally probably prefer field being passed to every function, rather than relying on the lifetime of your pointer being safe like you do now.